woocommerce/woocommerce-gateway-stripe
# Move charge related code to separate try-catch to prevent renewal failure

Reported in 8657329-zd-a8c 
Discussion p1726467962851179-slack-C7U3Y3VMY

In [process_subscription_payment](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L232C18-L232C46) function under one try-catch block, we make 2 API calls during the payment flow. 
1. The first API call is to [create and confirm the intent](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L312) which mainly processes the subscription payment. In this step, the card is already charged if the request is successful.
2. We make an API call to get the `charge` object to find the transaction ID. ([here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L373) and [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L403))

The merchant reported that their customers are charged multiple times for the same subscription renewal. They have the retry mechanism enabled for failed payments.
We have noticed from their logs that they encountered some timeouts during requests to Stripe. In the above function, when they encountered the timeouts in step 2, it marked their order as failed. As a result the retry mechanism was triggered and customer was charged again. 

## Changes proposed in this Pull Request:
- Moved the API requests under two separate try-catch blocks. 
- If the first request (create and confirm intent) fails, the card won't be charged. In the catch block of this section, we change the order status to `failed`.
- In the catch block of second API request (get charge), we only log the error but won't change the order status to failed as the card should be already charged.

## Testing instructions

- Checkout to `develop` branch.
- From **WooCommerce > Settings > Subscriptions**, enable `Retry Failed Payments`.
- As a shopper, purchase a subscription.
- In [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L401) add the following code to intentionally fail the get `charges` request.

```
// tweak charge id to fail intentionally
if ( ! empty( $response->latest_charge ) ) {
    $response->latest_charge = $response->latest_charge . 'm';
}
```
- As an admin, go to the subscription edit page and renew the subscription from `Order actions` dropdown.
- Go to the renewal order edit page.
- Notice that the order has failed but in the order notes you will see a payment intent was created.
- Go to your Stripe dashboard and you will find the corresponding transaction.
- Now checkout to this branch.
- Follow the above steps to reproduce the issue.
- This time the order should not fail but you should see the `charges` related error in your logs.
